File: src/my_project/medium_problems/from51to100/house_rober_ii_dp.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) ==1:
            return nums[0]
        
        return max(self.nocirle(nums[1:]), self.nocirle(nums[:-1]))
    
    def nocirle(self, nums):
        if not nums:
            return 0
        if len(nums) <= 2:
            return max(nums)
        
        dp = [0 for _ in range(len(nums))]
        dp[0] = nums[0]
        dp[1] = max(nums[0],nums[1])
        
        for i in range(2, len(nums)):
            dp[i] = max(dp[i-2] + nums[i], dp[i-1])
            
        return dp[-1]    

File: my_project_typescript/medium_problems_typescript/from1to50/houseRoberiiDP.ts
function rob(nums: number[]): number {
    if (nums.length == 1){
        return nums[0]
    }
    return Math.max(maxRob(nums.slice(1)), maxRob(nums.slice(0,nums.length-1)))
};

function maxRob(nums:number[]):number{
    if (!nums){
        return 0
    }
    if (nums.length <= 2){
        return Math.max.apply(null,nums)
    }
    let dp: Array<number> = Array(nums.length).fill(0)
    dp[0] = nums[0]
    dp[1] = Math.max(nums[0],nums[1])
    for (let i = 2; i < nums.length; i++){
        dp[i] = Math.max(dp[i-2]+nums[i], dp[i-1])
    }
    return dp[nums.length-1]
}